Prepare for latest Docker Engine: Add npipe

StefanScherer opened the pull request:
After testing https://github.com/docker/docker/pull/21040 I found out that I have to add another option `-H npipe:////./pipe/docker_engine` to the Docker Engine to listen on the named pipe as well. See https://github.com/docker/docker/pull/19911 for the introduction of the npipe feature.

This PR is incompatible with the TP4 Docker Engine and should be merged after a new Windows Docker Engine is released.

I just want to write it down somewhere if someone else is trying the latest Docker Engine.

enderb-ms commented:
@jhowardmsft

jhowardmsft commented on virtualization/windowscontainers/deployment/docker_windows.md:
```diff
@@ -42,11 +42,11 @@ mkdir %ProgramData%\docker
 if exist %certs%\server-cert.pem (goto :secure)
 
-docker daemon -D -b "Virtual Switch"
+docker daemon -D -b "Virtual Switch" -H npipe:////./pipe/docker_engine
```
This is not needed as the default is this.

jhowardmsft commented on virtualization/windowscontainers/management/container_networking.md:
````diff
@@ -124,7 +124,7 @@ Stop-Service docker
 
 ```powershell
-docker daemon -D -b “New Switch Name"
+docker daemon -D -b “New Switch Name" -H npipe:////./pipe/docker_engine
````
Shouldn't need a -H as this is the default.

StefanScherer commented on virtualization/windowscontainers/deployment/docker_windows.md:
````diff
@@ -121,7 +121,7 @@ Download docker.exe from `https://aka.ms/tp4/docker` and copy it to the `windows
 
 ```powershell
-PS C:\> start-process cmd "/k docker daemon -D -b <Switch Name> -H 0.0.0.0:2375”
+PS C:\> start-process cmd "/k docker daemon -D -b <Switch Name> -H npipe:////./pipe/docker_engine -H 0.0.0.0:2375”
````
Oh, just found "bad" typographic double quotes at the end of this line and some other places as well.

StefanScherer commented on virtualization/windowscontainers/deployment/docker_windows.md:
````diff
@@ -169,7 +169,7 @@ Now, on your machine, open a PowerShell or CMD session, and run the Docker comma
+For example, if you would like to see the available images:
 
 ```powershell
 .\docker.exe -H tcp://<ip address of Nano Server>:2375 images
````
I think here a remote setup is wanted as the client runs on your machine and the Nano Server is another machine with different IP address.

jhowardmsft replied:
That should still be fine - use npipe:////servername/pipe/docker_engine. There should not be any explicit 2375 listening on TCP regardless of nano/servercore, local or remote.

StefanScherer commented:
@jhowardmsft Thanks for the review. I've removed the uneccesary -H options. I've rebased the PR.

(My editor (Atom) found a lot of trailing white spaces and removed them.)

scooley commented:
@StefanScherer the whitespace thing is such a mixed bag.  Sometimes it's unnecessary, sometimes it's two spaces to make a line break rather than paragraph break...

Atom is great about preserving the double spaces and removing the single spaces.  In short.  DOn't worry about it

jhowardmsft commented on virtualization/windowscontainers/management/container_networking.md:
````diff
@@ -124,7 +124,7 @@ Stop-Service docker
 
 ```powershell
-docker daemon -D -b “New Switch Name"
+docker daemon -D -b "New Switch Name"
````
Good catch

StefanScherer replied:
Just CMD+SHIFT+F (or CTRL+SHIFT+F) with the `“` to search through the whole repo. Should be done for `”` as well.

jhowardmsft commented on windows-server-container-tools/Install-ContainerHost/Install-ContainerHost.ps1:
```diff
@@ -1168,12 +1168,12 @@ mkdir %ProgramData%\docker
+
 :secure
-docker daemon -D -b "$global:SwitchName" -H 0.0.0.0:2376 --tlsverify --tlscacert=%certs%\ca.pem --tlscert=%certs%\server-cert.pem --tlskey=%certs%\server-key.pem
+docker daemon -D -b "$global:SwitchName" -H npipe:////./pipe/docker_engine -H 0.0.0.0:2376 --tlsverify --tlscacert=%certs%\ca.pem --tlscert=%certs%\server-cert.pem --tlskey=%certs%\server-key.pem
```
Actually, thought about this some more. I think (not verified) that TLS will work over npipe, in which case the TCP listener shouldn't be required. Need someone like @taylorb-microsoft to verify if that is the case.

StefanScherer replied:
I've tested this with @taylorb-microsoft's TLS script today morning and it worked fine. TLS over 2376 from outside and probably no TLS with npipe to work within the TP4 machine locally. Tested this with a Mac Docker Client against the TP4 VM.

Works just like the Linux version:

```
/usr/local/bin/docker daemon -D -g /var/lib/docker -H unix:// -H tcp://0.0.0.0:2376 --label provider=vmwarefusion --tlsverify --tlscacert=/var/lib/boot2docker/ca.pem --tlscert=/var/lib/boot2docker/server.pem --tlskey=/var/lib/boot2docker/server-key.pem -s aufs
```

StefanScherer replied:
I will verify this again, tried this today morning and had a problem. That's why I search through the source and found the default pipe name.

jhowardmsft replied:
It does.

```
E:\go\src\github.com\docker\docker\docker>docker daemon -H npipe://
time="2016-03-10T13:46:14.892344900-08:00" level=info msg="Graph migration to co
ntent-addressability took 0.00 seconds"
time="2016-03-10T13:46:14.892843200-08:00" level=info msg="Loading containers: s
tart."

time="2016-03-10T13:46:14.892843200-08:00" level=info msg="Loading containers: d
one."
time="2016-03-10T13:46:14.893343400-08:00" level=info msg="Daemon has completed
initialization"
time="2016-03-10T13:46:14.893343400-08:00" level=info msg="Docker daemon" commit
=c545f17-dirty graphdriver=windowsfilter version=1.11.0-dev
time="2016-03-10T13:46:14.898841600-08:00" level=info msg="API listen on //./pip
e/docker_engine"
```

StefanScherer replied:
Confirmed. I'll simplify the PR.

taylorb-microsoft commented:
Ignoring the white space the raw change is the same change that I was playing with...

I think @mebersol was also planning to make this change and you beat him to it :+1:

mebersol commented:
Is this change backwards-compatible with TP4 Docker?

StefanScherer commented:
@mebersol No, this is **not compatible** and should be merged after a new release is there. Probably the download link should be updated to make this clear (changing the line with `$DockerPath = "https://aka.ms/tp4/docker",`).

Using TLS with the (Nov 23) TP4 docker.exe also is a little tricky. As the old default was port 2375 w/o TLS you can't just add `-H tcp://127.0.0.1:2375 -H tcp://0.0.0.0:2376` because then local port 2375 also is TLS. Docker Engine only turns on and off TLS globally for all tcp ports.

So TLS makes sense with the `npipe://` feature to work both locally w/o TLS and remote with TLS.

mebersol commented:
OK, thanks for the confirm - that's what I thought.

scooley commented:
@mebersol @jhowardmsft -- reading the thread I'm not sure if I should merge into live or merge into TP5 staging.

jhowardmsft commented:
TP5 staging

neilpeterson commented:
@jhowardmsft -  This is what is working for me, can you verify that this is the expected configuration  and behavior?

Start Docker (non-tls):
```
docker daemon -D -H npipe:// -H 0.0.0.0:2375
```

Access Docker from remote system:
```
docker -H tcp://10.121.106.78:2375
```

jhowardmsft commented:
All you're saying there is that the daemon is listening (securely) on npipe locally, and (insecurely) on all network interfaces over TCP.  I don't think we want to recommend insecure TCP configuration ever.

neilpeterson commented:
Thanks - fair enough. So then the recommendations would be:

```
docker daemon -D
```
and / or
```
docker daemon -D -H npipe:// -H 0.0.0.0:2376 --tlsverify --tlscacert=%certs%\ca.pem --tlscert=%certs%\server-cert.pem --tlskey=%certs%\server-key.pem
```

Just want to validate that the `-H npipe://` is necessary. From my insecure tests, it is.

Thanks

StefanScherer commented:
Yes. Having no -H option it defaults to npipe:// (as it would default to unix:// on Linux), and giving one -H for insecure/secure TCP connections you also need the second -H for npipe://.

One other detail. I compare the command lines with the Linux distros at the moment.
A boot2docker / docker-machine VM uses this command line:

```
/usr/local/bin/docker daemon -D -g /var/lib/docker -H unix:// -H tcp://0.0.0.0:2376 ...
``` 

So there is a `tcp://0.0.0.0:2376` instead of just `0.0.0.0:2376`.
Should we add the `tcp://` to be more consistent with the Linux docs?

neilpeterson commented:
Thanks Stefan, agreed on `tcp://`. I've just tested, it works fine, I will adopt this for TP5 ++ docs.

neilp

StefanScherer commented:
OK, I've fixed it in this PR as well.

friism commented:
@jhowardmsft I just tried tooling around with mounting nipe into a windows container using `-v`. If one can negotiate the path-wrangling, do you think that's possible?

Mounting the Docker socket is common (for reasons good and bad) on Linux

friism commented:
@taylorb-microsoft are you thinking about ways to make socket mounting happen?

jhowardmsft commented:
Not technically possible. Windows is not Linux!

friism commented:
Could we at least entertain workarounds...? We shouldn't bake it into the engine, but I wonder if there might be a way to proxy the HTTP API into a running container some other way.

friism commented:
Could we at least entertain workarounds...? We shouldn't bake it into the engine, but I wonder if there might be a way to proxy the HTTP API into a running container some other way.

jhowardmsft commented:
IF (capitals) there is, it's way down the list. If it requires platform support (it likely will), I am extremely doubtful there are resources to entertain working on this this side of RTM.

taylorb-microsoft commented:
I want to make sure I understand the scenario for socket mounting...  Is the scenario for getting basic network access in the container w/out using NAT?

friism commented:
@taylorb-microsoft no, it's mounting the docker engine API socket. That way you can run a docker container that can introspect itself and the other containers running on the host, and schedule new containers. Software like UCP uses this for installing itself.

friism commented:
( @taylorb-microsoft again, this is non-urgent - I'm just interested in exploring what we could do for this )

neilpeterson commented:
I believe these changes have been merged into today's release. I am going to close out this PR. If I've overlooked anything, just let me know.

StefanScherer commented:
No problem. Yes the new runDockerDaemon.cmd looks good.
